norse/torch/module/snn.py

```diff
@@ -0,0 +1,201 @@
+            self.p,
+            self.dt,
+        )
+
+
+class RecurrentSNN(torch.nn.Module):
+    """
+    The base module for recurrent spiking neural networks (RSNN) *with* time.
+
+    Parameters:
+        input_size (int): The number of input neurons
+        hidden_size (int): The number of hidden neurons
+        activation (RecurrentActivation): The activation function accepting an input tensor, state
+            tensor, input weights, recurrent weights, and parameters module, and returning a tuple
+            of (output spikes (one per timestep), state).
+        state_fallback (Callable[[torch.Tensor], torch.Tensor]): A function that can return a
+            default state with the correct dimensions, in case no state is provided in the
+            forward pass.
+        p (torch.nn.Module): The neuron parameters as a torch Module, which allows the module
+            to configure neuron parameters as optimizable.
+        input_weights (torch.Tensor): Weights used for input tensors. Defaults to a random
+            matrix normalized to the number of hidden neurons.
+        recurrent_weights (torch.Tensor): Weights used for input tensors. Defaults to a random
+            matrix normalized to the number of hidden neurons.
+        autopses (bool): Allow self-connections in the recurrence? Defaults to False. Will also
+            remove autopses in custom recurrent weights, if set above.
+        dt (float): Time step to use in integration. Defaults to 0.001.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
+        p: torch.nn.Module,
+        input_weights: Optional[torch.Tensor] = None,
+        recurrent_weights: Optional[torch.Tensor] = None,
+        autopses: bool = False,
+        dt: float = 0.001,
+    ):
+        super().__init__()
+        self.activation = activation
+        self.state_fallback = state_fallback
+        self.p = p
+        self.dt = dt
+
+        if input_weights is not None:
+            self.input_weights = input_weights
+        else:
+            self.input_weights = torch.nn.Parameter(
+                torch.randn(hidden_size, input_size) * torch.sqrt(2 / hidden_size)
+            )
+        if recurrent_weights is not None:
+            self.recurrent_weights = recurrent_weights
+        else:
+            self.recurrent_weights = torch.randn(hidden_size, hidden_size) * torch.sqrt(
+                2 / hidden_size
+            )
+        self.recurrent_weights = torch.nn.Parameter(
+            recurrent_weights if autopses else remove_autopses(recurrent_weights)
+        )
+
+    def forward(self, input_tensor: torch.Tensor, state: torch.Tensor):
+        state = state if state is not None else self.state_fallback(input_tensor)
+
+        T = input_tensor.shape[0]
+        outputs = []
+
+        for ts in range(T):
```

Comment:
I wondered about this the other day: you can either have a time loop per neuron/layer (as done here), or you can have only one for the entire network. In the former you would go through all time steps for a single layer and then move on to the next and go through all steps for that one etc.; in the latter you would only have one time loop, but for each iteration you would go through all layers.

I was wondering whether there would be a different outcome for these?

Reply:
I think we should facilitate both. And theoretically, the outcome should be the same. The first case is exactly taken care of by the above, but to do the second case, you'd need something that groups all the dynamics into a single module and runs them over all the timesteps. I think this is pretty much what @cpehle tried to do in the [Lift module](https://github.com/norse/norse/blob/master/norse/torch/module/lift.py#L57).

The question we should be asking ourselves, I guess, is whether there is a better way of achieving this. Should we, for instance, augment the [SequentialState](https://github.com/norse/norse/blob/master/norse/torch/module/sequential.py) to also support time? Can we think of something better? Or should we not include these abstractions at all and let people code this themselves?